### DISCUSSION TITLE
How to check if a valid container environment is available?

### DISCUSSION THREAD

**Discussion by tvandinther** [testcontainers/testcontainers-dotnet#1130] [Q&A] [answered]
I would like to skip tests instead of failing them if the environment does not have a valid environment for running the containers. Essentially a check on whether the socket is available. Is this something that the library supports with some static property somewhere? How else can I achieve this?

**Comment by HofmeisterAn** [comment]
There is nothing built in like this in Testcontainers. I am not quite sure if that is really the responsibility of Testcontainers, rather than the responsibility of the test framework (although we can provide features that help to determine whether tests should be skipped). But we do something similar in our tests to distinguish between container platforms. In xUnit.net, you can either implement a custom [FactAttribute](https://github.com/testcontainers/testcontainers-dotnet/blob/5c677d9ac043d414ecc3c165c265700c98668808/tests/Testcontainers.Platform.Windows.Tests/SkipOnLinuxEngineAttribute.cs) or use a [Trait](https://github.com/testcontainers/testcontainers-dotnet/blob/5c677d9ac043d414ecc3c165c265700c98668808/tests/Testcontainers.PostgreSql.Tests/PostgreSqlContainerTest.cs#L18) annotation. Then you can use the [--filter](https://learn.microsoft.com/dotnet/core/tools/dotnet-test#filter-option-details) flag to distinguish between the tests. Does that help?

**Reply by tvandinther** [reply]
Thanks for the response. I currently use the SkippableFact to skip tests if the container creation fixture detects a. Non-linux OS. I was hoping to be a bit more descriptive in that check and specifically target whether the container environment is valid.

I am also on the fence on whether this is a good pattern in the first place as it might accidentally hide the tests on Linux if the environment is misconfigured to fail the condition check.

I'd still be curious what the best way to perform the check would be. Would you wrap the container creation in a try catch and skip on exception? I can see this being problematic for the reason stated above though.

**Reply by HofmeisterAn** [reply, marked as answer]
> if the container creation fixture detects a. Non-linux OS.

What do you mean by that? The host OS, the contianer OS, the enabled engine (Linux/Windows)?

> Would you wrap the container creation in a try catch and skip on exception?

I am not sure if I understood what you are trying to achieve (maybe you need to elaborate on that), but if you would like to run tests depending on the supported container OS (enabled Docker engine), you can use a similar approach to ours. We check the `docker version --format '{{.Server.Os}}'` response to determine whether the container runtime (host) supports Linux or Windows containers.

https://github.com/testcontainers/testcontainers-dotnet/blob/c0932f2170cf53d15a69a52626b736606a4d28f8/tests/Testcontainers.Commons/DockerCli.cs#L46-L47

**Reply by tvandinther** [reply]
Thanks, that looks like something that will help.
